scalanet/src/io/iohk/scalanet/peergroup/udp/StaticUDPPeerGroup.scala

```diff
@@ -0,0 +1,444 @@
+                }
+              }
+            })
+
+          ()
+        }
+      })
+      .bind(localAddress)
+
+  // Wait until the server is bound.
+  private def initialize(): Task[Unit] =
+    for {
+      _ <- raiseIfShutdown
+      _ <- toTask(serverBinding).onErrorRecoverWith {
+        case NonFatal(ex) =>
+          Task.raiseError(InitializationError(ex.getMessage, ex.getCause))
+      }
+      _ <- Task(logger.info(s"Server bound to address ${config.bindAddress}"))
+    } yield ()
+
+  private def shutdown(): Task[Unit] = {
+    for {
+      _ <- Task(logger.info(s"Shutting down UDP peer group for peer ${config.processAddress}"))
+      // Mark the group as shutting down to stop accepting incoming connections.
+      _ <- isShutdownRef.set(true)
+      _ <- Task(serverSubject.onComplete())
+      // Release client channels.
+      _ <- clientChannelsRef.get.map(_.values.flatten.toList.map(_._2.attempt).sequence)
+      // Release server channels.
+      _ <- serverChannelsRef.get.map(_.values.toList.map(_._2.attempt).sequence)
+      // Stop the in and outgoing traffic.
+      _ <- toTask(serverBinding.channel.close())
+    } yield ()
+  }
+
+}
+
+object StaticUDPPeerGroup extends StrictLogging {
+  case class Config(
+      bindAddress: InetSocketAddress,
+      processAddress: InetMultiAddress
+  )
+  object Config {
+    def apply(bindAddress: InetSocketAddress): Config = Config(bindAddress, InetMultiAddress(bindAddress))
+  }
+
+  private type ChannelAlloc[M] = (ChannelImpl[M], Release)
+
+  def apply[M: Codec](config: Config)(implicit scheduler: Scheduler): Resource[Task, StaticUDPPeerGroup[M]] =
+    makeEventLoop.flatMap { workerGroup =>
+      Resource.make {
+        for {
+          isShutdownRef <- Ref[Task].of(false)
+          clientChannelsRef <- Ref[Task].of(Map.empty[InetSocketAddress, Set[ChannelAlloc[M]]])
+          serverChannelsRef <- Ref[Task].of(Map.empty[InetSocketAddress, ChannelAlloc[M]])
+          serverSubject = ConnectableSubject[ServerEvent[InetMultiAddress, M]]()
+          peerGroup = new StaticUDPPeerGroup[M](
+            config,
+            workerGroup,
+            isShutdownRef,
+            serverSubject,
+            serverChannelsRef,
+            clientChannelsRef
+          )
+          _ <- peerGroup.initialize()
+        } yield peerGroup
+      }(_.shutdown())
+    }
+
+  // Separate resource so if the server initialization fails, this still gets shut down.
+  private val makeEventLoop =
+    Resource.make {
+      Task(new NioEventLoopGroup())
+    } { group =>
+      toTask(group.shutdownGracefully())
+    }
+
+  private class ChannelImpl[M](
+      nettyChannel: io.netty.channel.Channel,
+      localAddress: InetSocketAddress,
+      remoteAddress: InetSocketAddress,
+      messageSubject: ConnectableSubject[ChannelEvent[M]],
+      // Prevent race conditions between closing the channel and publishing messages.
+      subjectSemaphore: Semaphore[Task],
```

Comment:
I wonder if we could get rid of this semaphore by using closing scheme similar to `DynamicUDPPeerGroup` i.e 
closing only  completes netty promise, and listener (which is also working on netty event loop) removes channel from map. Then there is no racing invloved as the thread which intercept the message, and closes channel is the same thread.

Reply:
I see what you mean you'd want the `onCompleted` to be called from somewhere else. I don't really want to complicate it for two reasons: the semaphore isn't blocking so it's probably not slowing anything down, and the next PR removes the `ConnectableSubject` anyway so we won't have this problem. It's also likely that the Observables protect against this too, so this is just being overly cautions, but adding netty promises to the mix sounds complicated.

Reply:
hmm that when i do not agree as:
- using netty concurrency primitives in not complicating, as rather natural way to interface with netty. I like the monix approach to this things i.e dirty inside and clean interface outside.
- any maybe overhead of semaphore is not much, but it is on lowest level of the stack. 
I do not want to block rest of the work, but imo after removing `ConnectableSubject` we should maybe revisit is.

Reply:
I agree that there's room to optimise and I appreciate your suggestions a lot, it would just probably take me too long to figure it out now. I know we can do a completely different implementation of this whole thing and now there are a few quick and dirty `runSyncUnsafe` instances which I hate. I do find the cats-effect machinery easier to reason about though, and I hope that on the long run we can find better integration patterns. FWIW the semaphore is removed in the next PR, and I changed this one to do the message handling asynchronously.